From c0bda09fbdbde46dfb69794557fa57f4eb799014 Mon Sep 17 00:00:00 2001
From: Zhipeng Xie <xiezhipeng1@huawei.com>
Date: Thu, 20 Jun 2019 12:05:16 +0800
Subject: [PATCH 3/3] livepatch: optimize module symbol lookup

module_kallsyms_on_each_symbol traverse every symbol in each module.
When we known the object module, we just traverse symbols in the
object module symbol list. we can reduce patch load time after apply
this patch.

Signed-off-by: Zhipeng Xie <xiezhipeng1@huawei.com>
---
 include/linux/module.h   |  6 ++++++
 kernel/livepatch/Kconfig | 10 ++++++++++
 kernel/livepatch/core.c  |  4 ++++
 kernel/module.c          | 38 ++++++++++++++++++++++++++++++++++++++
 4 files changed, 58 insertions(+)

diff --git a/include/linux/module.h b/include/linux/module.h
index b2b2c742a..86b2a4ea3 100644
--- a/include/linux/module.h
+++ b/include/linux/module.h
@@ -642,6 +642,12 @@ int module_kallsyms_on_each_symbol(int (*fn)(void *, const char *,
 					     struct module *, unsigned long),
 				   void *data);
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_LOAD_TIME_OPTIMIZATION
+int object_module_on_each_symbol(const char *objname, int (*fn)(void *,
+					const char *, struct module *,
+					unsigned long), void *data);
+#endif
+
 extern void __noreturn __module_put_and_exit(struct module *mod,
 			long code);
 #define module_put_and_exit(code) __module_put_and_exit(THIS_MODULE, code)
diff --git a/kernel/livepatch/Kconfig b/kernel/livepatch/Kconfig
index 401cd2a93..c0de6bb67 100644
--- a/kernel/livepatch/Kconfig
+++ b/kernel/livepatch/Kconfig
@@ -131,5 +131,15 @@ config EULEROS_LIVEPATCH_LONG_SYMBOL
           .klp.sym.objname-ref_name,ref_offset to find a unique symbol
           and add ref_offset to get the long name symbol address.
 
+config EULEROS_LIVEPATCH_LOAD_TIME_OPTIMIZATION
+        bool "patch load time optimization"
+        depends on LIVEPATCH
+        default y
+        help
+          module_kallsyms_on_each_symbol traverse every symbol in each module.
+          When we known the object module, we just traverse symbols in the
+          object module symbol list. we can reduce patch load time after apply
+          this patch.
+
 endmenu
 endif
diff --git a/kernel/livepatch/core.c b/kernel/livepatch/core.c
index e8a28e0a3..fb29d6b16 100644
--- a/kernel/livepatch/core.c
+++ b/kernel/livepatch/core.c
@@ -235,7 +235,11 @@ static int klp_find_object_symbol(const char *objname, const char *name,
 
 	mutex_lock(&module_mutex);
 	if (objname)
+#ifdef CONFIG_EULEROS_LIVEPATCH_LOAD_TIME_OPTIMIZATION
+		object_module_on_each_symbol(objname, klp_find_callback, &args);
+#else
 		module_kallsyms_on_each_symbol(klp_find_callback, &args);
+#endif
 	else
 		kallsyms_on_each_symbol(klp_find_callback, &args);
 	mutex_unlock(&module_mutex);
diff --git a/kernel/module.c b/kernel/module.c
index 23c5349a5..6b6fc3587 100644
--- a/kernel/module.c
+++ b/kernel/module.c
@@ -4506,6 +4506,44 @@ int module_kallsyms_on_each_symbol(int (*fn)(void *, const char *,
 	}
 	return 0;
 }
+
+#ifdef CONFIG_EULEROS_LIVEPATCH_LOAD_TIME_OPTIMIZATION
+int object_module_on_each_symbol(const char *objname, int (*fn)(void *,
+				 const char *, struct module *,
+				 unsigned long), void *data)
+{
+	struct module *mod;
+	unsigned int i;
+	int ret;
+	struct mod_kallsyms *kallsyms;
+	int found = 0;
+
+	list_for_each_entry(mod, &modules, list) {
+		if (mod->state == MODULE_STATE_UNFORMED)
+			continue;
+
+		if (!strcmp(objname, mod->name)) {
+			found = 1;
+			break;
+		}
+	}
+	if (!found)
+		return -EINVAL;
+
+	/* We hold module_mutex: no need for rcu_dereference_sched */
+	kallsyms = mod->kallsyms;
+
+	for (i = 0; i < kallsyms->num_symtab; i++) {
+		ret = fn(data, kallsyms->strtab + kallsyms->symtab[i].st_name,
+			 mod, kallsyms->symtab[i].st_value);
+		if (ret != 0)
+			return ret;
+	}
+
+	return 0;
+}
+#endif
+
 #endif /* CONFIG_KALLSYMS */
 
 /* Maximum number of characters written by module_flags() */
-- 
2.27.0

